neat_periodic_task/lib/neat_periodic_task.dart

```diff
@@ -0,0 +1,337 @@
+    }
+  }
+
+  Future<NeatTaskStatus> _getStatus() async {
+    final data = await _statusProvider.get();
+    return NeatTaskStatus.deserialize(data);
+  }
+
+  /// Sleep, but stop sleeping if we are stopping.
+  Future<void> _sleep(Duration delay) => Future.any([
+        Future.delayed(delay),
+        _stopping.future,
+      ]);
+
+  /// Do an iteration where we check the status and either sleep or attempt to
+  /// claim the task and run.
+  Future<void> _iteration() async {
+    _log.finest(() => 'fetching status for "$_name"');
+    final status = await _getStatus();
+    if (status.version > NeatTaskStatus.currentVersion) {
+      _log.warning(
+          'status for "$_name" with newer version found, version: ${status.version}');
+      await _sleep(_interval);
+      return;
+    }
+
+    // Find time elapsed since last status update.
+    final now = DateTime.now().toUtc();
+    final elapsed = now.difference(status.updated);
+    _log.finest(
+        () => 'time elapsed since status for "$_name" was written $elapsed');
+
+    // If state is 'finished' the delay before next run is _interval, otherwise
+    // we assume state us 'running' as delay is only _timeout.
+    var delay = _interval;
+    if (status.state != 'finished') {
+      delay = _timeout;
+    }
+
+    // If delay isn't past yet, we sleep.
+    if (elapsed < delay) {
+      Duration d = delay ~/ 2;
+      // Always sleep at least minCycle to ensure the iteration doesn't spin too
+      // fast as we approach the next iteration.
+      if (d < _minCycle) {
+        d = _minCycle;
+      }
+      // Never sleep more than maxCycle, as we must wake-up and print a log line
+      // that says we've checked the status of the task. Operators can either
+      // monitor the message saying this was done, or they can monitor the
+      // message saying that the status was monitored.
+      if (d > _maxCycle) {
+        d = _maxCycle;
+      }
+      if (status.state == 'finished') {
+        _log.info('### [ALIVE] neat-periodic-task: "$_name"');
+      }
+      _log.finest(() => 'NeatPeriodicTaskScheduler "$_name" sleeps $d');
+      await _sleep(d);
+
+      // Return such that we do another _iteration() call..
+      return;
+    }
+
+    // If elapsed >= delay, then we claim and run the task.
+    await _claimAndRun(status.update(
+      owner: Slugid.nice().toString(),
+      state: 'running',
+      updated: now,
+    ));
+  }
+
+  /// Claim the status and attempt to run.
+  Future<void> _claimAndRun(NeatTaskStatus status) async {
+    // Attempt to claim the task, if we're not the owner we failed to stop.
+    _log.finest(() => 'Attempt to claim task status for "$_name"');
+    final claimed = await _statusProvider.set(status.serialize());
+    if (!claimed) {
+      _log.finest(() => 'Failed to claim task status for "$_name"');
+      return;
+    }
+
+    try {
+      _log.info('### [START] neat-periodic-task: "$_name"');
+      await _task().timeout(_timeout);
```

Comment:
You could upgrade the Task return-type with a cancel.

Reply:
Agree... it would be a neat feature.. But outside the scope for now.